File: haprad2/exclusive_model.f
      SUBROUTINE exclusive_model(q2m, wm, csthcm, st, sl,
     &                           stt, stl, stlp)
      IMPLICIT NONE

      DOUBLE PRECISION st, sl, stt, stl, stlp
      INTEGER nq, nw, nt, iq, iw, it, nc, narg(3)
      PARAMETER(nq = 18)
      PARAMETER(nw = 47)
      PARAMETER(nt = 61)

      DOUBLE PRECISION q2, w, csthcm, dfint, ee, th_cm, degrad 
      DOUBLE PRECISION q2_pn(nq), w_pn(nw), th_cm_pn(nt)
      DOUBLE PRECISION ft_cs(nq, nw, nt), fl_cs(nq, nw, nt)
      DOUBLE PRECISION ftt_cs(nq, nw, nt)
      DOUBLE PRECISION ftl_cs(nq, nw, nt), ftlp_cs(nq, nw, nt)
      DOUBLE PRECISION arg(3), rarg(1000), a2, a30, a31, a3, wcor, q2cor
      DOUBLE PRECISION q2m, wm

      DATA q2_pn/0.0d0, 0.3d0, 0.6d0, 0.9d0, 1.2d0, 1.5d0, 1.8d0, 2.1d0,
     &           2.4d0, 2.7d0, 3.0d0, 3.3d0, 3.6d0, 3.9d0, 4.2d0, 4.5d0,
     &           4.8d0, 5.0d0/
      DATA w_pn/1.08d0, 1.10d0, 1.12d0, 1.14d0, 1.16d0, 1.18d0, 1.20d0,
     &          1.22d0, 1.24d0, 1.26d0, 1.28d0, 1.30d0, 1.32d0, 1.34d0,
     &          1.36d0, 1.38d0, 1.40d0, 1.42d0, 1.44d0, 1.46d0, 1.48d0,
     &          1.50d0, 1.52d0, 1.54d0, 1.56d0, 1.58d0, 1.60d0, 1.62d0,
     &          1.64d0, 1.66d0, 1.68d0, 1.70d0, 1.72d0, 1.74d0, 1.76d0,
     &          1.78d0, 1.80d0, 1.82d0, 1.84d0, 1.86d0, 1.88d0, 1.90d0,
     &          1.92d0, 1.94d0, 1.96d0, 1.98d0, 2.00d0/
      DATA th_cm_pn/0.d0, 3.d0, 6.d0, 9.d0, 12.d0, 15.d0, 18.d0, 21.d0,
     &              24.d0, 27.d0, 30.d0, 33.d0, 36.d0, 39.d0, 42.d0,
     &              45.d0, 48.d0, 51.d0, 54.d0, 57.d0, 60.d0, 63.d0,
     &              66.d0, 69.d0, 72.d0, 75.d0, 78.d0, 81.d0, 84.d0,
     &              87.d0, 90.d0, 93.d0, 96.d0, 99.d0, 102.d0, 105.d0,
     &              108.d0, 111.d0, 114.d0, 117.d0, 120.d0, 123.d0,
     &              126.d0, 129.d0, 132.d0, 135.d0, 138.d0, 141.d0,
     &              144.d0, 147.d0, 150.d0, 153.d0, 156.d0, 159.d0,
     &              162.d0, 165.d0, 168.d0, 171.d0, 174.d0, 177.d0,
     &              180.d0/
      DATA nc/0/, narg/nq, nw, nt/, degrad/57.29577952d0/, a2/1.15d0/
      DATA a30/-1.23d0/, a31/0.16d0/

      COMMON /exlusive/ rarg, ft_cs, fl_cs, ftt_cs, ftl_cs, ftlp_cs

C Init
      st = 0.d0
      sl = 0.d0
      stt = 0.d0
      stl = 0.d0
      stlp = 0.d0

C new variables
      q2 = q2m
      w = wm
      th_cm = acos(csthcm)*degrad

C Check kinematics
      IF (q2 .LT. 0.0) THEN
         PRINT *, 'Warning: Q2<0 in exclusive model!'
         PRINT *, 'Using Q2 = 0'
         q2 = 0.d0
      ENDIF

      IF(q2 .GT. 5.0) THEN
C        PRINT *, 'Warning: Q2>5 GeV^2 in exclusive model!'
C        PRINT *, 'Using extrapolation from MAID2003'
         q2cor = (5.d0 ** a2) / (q2 ** a2)
         q2 = 5.d0
      ELSE
         q2cor = 1.d0
      ENDIF

      IF(w .LT. 1.07784) RETURN

      IF(w .GT. 2.0) THEN
C        PRINT *, 'Warning: W>2 GeV in exclusive model!'
C        PRINT *, 'Using extrapolation from MAID2003 (A.Browman PRL35, Cornell)'
         a3 = a30 + a31 * th_cm
         IF (th_cm .LT. 50.0) a3 = a30 + a31 * 50.d0
         IF (th_cm .GT. 100.0) a3 = a30 + a31 * 100.d0
         wcor = (2.d0 ** a3) / (w ** a3)
         w = 2.d0
      ELSE
         wcor = 1.d0
      ENDIF

      IF (ABS(csthcm) .GT. 1.0) RETURN

C Read data from file
      IF (nc .EQ. 0) THEN
         OPEN (44, FILE = 'pi_n_maid.dat', STATUS = 'old')

         DO iq = 1, nq
            DO iw = 1, nw
               DO it = 1, nt
                  READ (44, *) ft_cs(iq, iw, it), fl_cs(iq, iw, it),
     &                         ftt_cs(iq, iw, it), ftl_cs(iq, iw, it),
     &                         ftlp_cs(iq, iw, it)
               ENDDO
            ENDDO
         ENDDO

         CLOSE(44)

         DO iq = 1, nq
            rarg(iq) = q2_pn(iq)
         ENDDO

         DO iw = 1, nw
            rarg(iw+nq) = w_pn(iw)
         ENDDO

         DO it = 1, nt
            rarg(it+nw+nq) = th_cm_pn(it)
         ENDDO

         nc = nc + 1
      ENDIF

C Extract interpolated cross section
C     if (q2 .GE. q2_pn(1) .AND.
C    &         w .GE. w_pn(1) .AND.
C    &         th_cm .GE. th_cm_pn(1) .AND.
C    &         q2 .LE. q2_pn(nq) .AND.
C    &         w .LE. w_pn(nw) .AND.
C    &         th_cm .LE. th_cm_pn(nt)) THEN
         arg(1) = q2
         arg(2) = w
         arg(3) = th_cm

         st = dfint(3, arg, narg, rarg, ft_cs) * wcor * q2cor
         sl = dfint(3, arg, narg, rarg, fl_cs) * wcor * q2cor
         stt = dfint(3, arg, narg, rarg, ftt_cs) * wcor * q2cor
         stl = 2.d0 * dfint(3, arg, narg, rarg, ftl_cs) * wcor * q2cor
         stlp = 0.d0
C        stlp = dfint(3, arg, narg, rarg, ftlp_cs) * wcor * q2cor
C     ELSE
C        st = 0d0
C        sl = 0d0
C        stt = 0d0
C        stl = 0d0
C        stlp = 0d0
C     ENDIF
      RETURN
      END

File: src/exclusive_model.cxx
#include <iostream>
#include <fstream.h>
#include <string>
#include "TMath.h"
#include "TRoot.h"

#define N 100000;

void exclusive_model(Double_t q2m, Double_t wm, Double_t csthcm, Double_t st, Double_t sl, Double_t stt, Double_t stl, Double_t stlp)
{
    Double_t nq = 18;
    Double_t nw = 47;
    Double_t nt = 61;

    Double_t q2_pn[nq] = {0.0, 0.3, 0.6, 0.9, 1.2, 1.5, 1.8,
                          2.1, 2.4, 2.7, 3.0, 3.3, 3.6, 3.9,
                          4.2, 4.5, 4.8, 5.0
                         };

    Double_t w_pn[nw] = {1.08, 1.10, 1.12, 1.14, 1.16, 1.18,
                         1.20, 1.22, 1.24, 1.26, 1.28, 1.30,
                         1.32, 1.34, 1.36, 1.38, 1.40, 1.42,
                         1.44, 1.46, 1.48, 1.50, 1.52, 1.54,
                         1.56, 1.58, 1.60, 1.62, 1.64, 1.66,
                         1.68, 1.70, 1.72, 1.74, 1.76, 1.78,
                         1.80, 1.82, 1.84, 1.86, 1.88, 1.90,
                         1.92, 1.94, 1.96, 1.98, 2.00
                        };
    Double_t th_cm_pn[nt] = {0., 3., 6., 9., 12., 15., 18., 21.,
                             24., 27., 30., 33., 36., 39., 42.,
                             45., 48., 51., 54., 57., 60., 63.,
                             66., 69., 72., 75., 78., 81., 84.,
                             87., 90., 93., 96., 99., 102., 105.,
                             108., 111., 114., 117., 120., 123.,
                             126., 129., 132., 135., 138., 141.,
                             144., 147., 150., 153., 156., 159.,
                             162., 165., 168., 171., 174., 177.,
                             180.
                            };
    Int_t narg[3] = {nq, nw, nt};
    Double_t nc = 0;
    Double_t degrad = 57.29577952;
    Double_t a2 = 1.15;
    Double_t a30 = -1.23;
    Double_t a31 = 0.16;
// Init
    st = 0.0;
    sl = 0.0;
    stt = 0.0;
    stl = 0.0;
    stlp = 0.0;

//new variables
    q2 = q2m;
    w = wm;
    th_cm = TMath::ACos(csthcm) * degrad;
//Check Kinematics
    if (q2 < 0.0) {
        std::cout << "Warning: Q2 < 0 in exclusive model!" << std::endl;
        std::cout << " Using Q2 = 0" << std::endl;
        q2 = 0;

    }
    if (q2 > 5) {
        q2cor = TMath::Power(5., a2) / TMath::Power(q2, a2);
        q2 = 5.;

    } else q2cor = 1.;

    if (w < 1.07784) return;

    if (w > 2) {
        if (th_cm < 50.0) a3 = a30 + a31 * 50.;
        if (th_cm > 100.0) a3 = a30 + a31 * 100.;

        a3 = a30 + a31 * th_cm;
        wcor = TMath::Power(2., a3) / TMath::Power(w, a3);
        w = 2.;

    } else wcor = 1.;

    if (TMath::Abs(csthcm) > 1) return;

    Double_t rarg[nq + nw + nt];
    Double_t ft_cs[N], fl_cs[N], ftt_cs[N], ftl_cs[N], ftlp_cs[N];

    if (nc == 0) {
        Int_t i = 0;
        FILE *input = fopen("pi_n_maid.dat", "r");

        while (fscanf(input, "%f   %f   %f   %f   %f", &ft_cs[i], &fl_cs[i], &ftt_cs[i], &ftl_cs, &ftlp_cs[i])) {
            i++;
        }
		fclose(input);

        for (Int_t i = 0; i < nq; i++)
            rarg[i] = q2_pn[i];

        for (Int_t j = 0; j < nw; j++)
            rarg[j + nq] = w_pn[j];

        for (Int_t k = 0; k < nt; k++)
            rarg[k + nq + nw] = th_cm_pn[k];

        nc++;

    }

    arg[0] = q2;
    arg[1] = w;
    arg[2] = th_cm;

    st = dfint(3, arg, narg, rarg, ft_cs) * wcor * q2cor;
    sl = dfint(3, arg, narg, rarg, fl_cs) * wcor * q2cor;
    stt = dfint(3, arg, narg, rarg, ftt_cs) * wcor * q2cor;
    stl = 2. * dfint(3, arg, narg, rarg, ftl_cs) * wcor * q2cor;
    stlp = 0;

	return;
}
